npm/podController.go

```diff
@@ -0,0 +1,627 @@
+		// Pod resource to be synced.
+		// (TODO): may consider using "c.queue.AddAfter(key, *requeueAfter)" according to error type later
+		if err := c.syncPod(key); err != nil {
+			// Put the item back on the workqueue to handle any transient errors.
+			c.workqueue.AddRateLimited(key)
+			return fmt.Errorf("error syncing '%s': %s, requeuing", key, err.Error())
+		}
+		// Finally, if no error occurs we Forget this item so it does not
+		// get queued again until another change happens.
+		c.workqueue.Forget(obj)
+		log.Logf("Successfully synced '%s'", key)
+		return nil
+	}(obj)
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+// syncPod compares the actual state with the desired, and attempts to converge the two.
+func (c *podController) syncPod(key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	// Get the Pod resource with this namespace/name
+	pod, err := c.podLister.Pods(namespace).Get(name)
+	// (TODO): Reduce scope of lock later
+	c.npMgr.Lock()
+	defer c.npMgr.Unlock()
+	if err != nil {
+		if errors.IsNotFound(err) {
+			log.Logf("pod %s not found, may be it is deleted", key)
+			// Find the npmPod object from PortMap local cache and start cleaning up processes
+			cachedNpmPodObj, exist := c.npMgr.PodMap[key]
+			if exist {
+				err = c.cleanUpDeletedPod(cachedNpmPodObj.getPodObjFromNpmPodObj())
+				if err != nil {
+					metrics.SendErrorLogAndMetric(util.PodID, "Error: %v when pod is not found", err)
+					return fmt.Errorf("Error: %v when pod is not found\n", err)
+				}
+			}
+		}
+		return err
+	}
+
+	if pod.DeletionTimestamp != nil || pod.DeletionGracePeriodSeconds != nil {
+		err = c.cleanUpDeletedPod(pod)
+		if err != nil {
+			metrics.SendErrorLogAndMetric(util.PodID, "Error: %v when DeletionTimestamp or DeletionGracePeriodSeconds is not nil", err)
+			return fmt.Errorf("Error: %v when DeletionTimestamp or DeletionGracePeriodSeconds is not nil\n", err)
+		}
+		return nil
+	}
+
+	err = c.syncAddAndUpdatePod(pod)
+	if err != nil {
+		metrics.SendErrorLogAndMetric(util.PodID, "Error: Failed to sync pod due to %v", err)
+		return fmt.Errorf("Failed to sync pod due to %v\n", err)
+	}
+
+	return nil
+}
+
+func (c *podController) syncAddedPod(podObj *corev1.Pod) error {
+	npmPodObj := newNpmPod(podObj)
+	podNs := util.GetNSNameWithPrefix(podObj.Namespace)
+	podKey, _ := cache.MetaNamespaceKeyFunc(podObj)
+	ipsMgr := c.npMgr.NsMap[util.KubeAllNamespacesFlag].IpsMgr
+	log.Logf("POD CREATING: [%s%s/%s/%s%+v%s]", npmPodObj.PodUID, podNs, npmPodObj.Name, npmPodObj.NodeName, npmPodObj.Labels, npmPodObj.PodIP)
+
+	// Add pod namespace if it doesn't exist
+	var err error
+	if _, exists := c.npMgr.NsMap[podNs]; !exists {
+		// (TODO): Any chance to return error?
```

Comment:
we should handle the error here

Reply:
This comment is about "newNs()", not c.npMgr.NsMap. newNs function always return nil. So it may be better to remove "return error" in newNs() function.